.github/workflows/test.yml

```diff
@@ -41,29 +41,29 @@ jobs:
     steps:
       - uses: actions/checkout@v4
 
-      - uses: actions/setup-python@v5
+      - uses: astral-sh/setup-uv@v3
         with:
-          python-version-file: ".python-version"
-          cache: "pip"
-          cache-dependency-path: "**/pyproject.toml"
-
-      - name: Install dependencies
-        run: |
-          python -m pip install --upgrade pip
-          pip install hatch
+          version: ">=0.4.0"
+          enable-cache: true
+          cache-dependency-glob: "**/pyproject.toml"
 
       - name: Build and lint Widget
         run: |
           cd packages/widget
-          hatch build
-          hatch fmt --check
+
+          uv build
+          uv run ruff check
+          uv run ruff format --check
 
       - name: Build, lint, and test Server
         run: |
           cd packages/duckdb-server
-          hatch build
-          hatch fmt --check
-          hatch run test:cov
+
+          uv build
+          uv run ruff check
+          uv run ruff format --check
+          uv run mypy
+          uv run --with pytest-cov pytest --cov-report=term-missing --color=yes --cov=pkg
```

Comment:
Depending on how you want to split out these, I believe you can run:

```
uv run ruff check
uv run ruff format --check
```

from the root of the repo. 

I  usally have a separate workflow for "Python Lint" e.g.,:


```sh
 python_lint:
    name: Lint Python
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - uses: astral-sh/setup-uv@v3
        with:
          python-version-file: ".python-version"
          cache: "pip"
          cache-dependency-path: "**/pyproject.toml"

      - name: Ruff
        run: |
          uv run ruff check
          uv run ruff format --check
          
  python_test:
    name: Test Python
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - uses: astral-sh/setup-uv@v3
        with:
          python-version-file: ".python-version"
          cache: "pip"
          cache-dependency-path: "**/pyproject.toml"

      - name: Widget
        run: |
          cd packages/widget
          uv build

      - name: Server
        run: |
          cd packages/duckdb-server
          uv run mypy
          uv run --with pytest-cov pytest --cov-report=term-missing --color=yes --cov=pkg
```

Reply:
I just moved out up. I didn't set up a separate workflow since the python one is plenty fast. 

setup-uv doesn't seem to like the python version file. 

<img width="515" alt="Screenshot 2024-10-31 at 10 43 57" src="https://github.com/user-attachments/assets/cd77d1d1-d7c1-4821-8de7-3361a86e68ea">

Reply:
> setup-uv doesn't seem to like the python version file.

Yeah, uv is a python installer. It will automatically use `.python-version` if it's in the root of the project when running commands. YOU can specify the version with `UV_PYTHON=` in the env if you want to test multple python versions in matrix.

Reply:
I think you could `setup-python` with a specific version and then `setup-uv` and run uv commands with `--system` to use the "system python" (i.e., whatever `setup-python` did), but i find the uv-only workflow most simple.

Reply:
https://github.com/manzt/anywidget/blob/478e742eaaf85d8fab68b006210680232863a488/.github/workflows/ci.yml#L56-L69

For example, I should change this to:

```yaml
  TestPython:
    name: Python / Test
    runs-on: ubuntu-latest
    strategy:
      matrix:
        python-version:
          - "3.8"
          - "3.9"
          - "3.10"
          - "3.11"
          - "3.12"
    steps:
      - uses: actions/checkout@v4
      - uses: pnpm/action-setup@v4
        with:
          run_install: true
      - uses: astral-sh/setup-uv@v3
        with:
          version: "0.4.x"
      - name: Run tests
        run: uv run --with pytest-cov pytest ./tests --color=yes --cov anywidget --cov-report xml
        env:
          UV_PYTHON: ${{ matrix.python-version }}
 ```